text/0002-new-architecture.md

```diff
@@ -0,0 +1,243 @@
+the situation that it is quite easy to ship regressions because the changes in `getsentry` were not considered.
+There is likely a whole range of code that does not need to live in `getsentry` and moving it to `sentry` would
+make the testsuite more reliable, faster to run and catch more regressions early.
+
+### Remove Pickle in Database
+
+We still use pickled models in database code quite extensively.  This reliance on pickle makes it harder than
+necessary to read and write from this outside of the main Python codebase.  Moving this to JSON will simplify the
+handling of moving such code into environments where other code wants to access it.  We already require depickling
+in the data pipeline for such cases.
+
+### Phase out Buffers where Possible
+
+Buffers as they exist today could in many cases be replaced by better uses of Kafka and clickhouse.  Some buffers are
+not even as important any more for the product.  For instance the total counts for issues are already hidden from the
+more prominent places in the UI.  The issue with buffers is that they are rather hard to scale, require the use of
+pickle for the model updates and are hard to work with when it comes to filtering.  For instance we are able to give
+you the total event count on an issue, but not when broken down by some of the filters that the UI wants to provide
+(environment and release is possible, but any tag search will not able to give the right counts).
+
+## Relay
+
+Within relay some changes are likely to be necessary for continued to support of providing a stable service.  In
+particular as more and more work moves into Relay certain risks that the current architecture is carrying need to
+be addressed.  The goals here are horizontal scalability, the ability to route data more efficiently and to recover
+from catastrophic failures.
+
+### Traffic Steering
+
+As we are now performing aggregations in Relay it's beneficial to be able to route traffic intelligently through the
+layers.  For instance we can achieve much better aggregations by ensuring that data from related metrics keys are
+forwarded to the same processing relays.
+
+### Partial Project Options
+
+Relays currently get one project config per project both via the HTTP protocol and by picking them up internally from
+a redis key.  The size of this config package is exploding as more and more functionality is added.  There are many
+benefits from splitting this up which should improve our ability to partially degrade service rather than the entire
+project, reduce the pressure of redis and make updates faster.  For instance an experimental feature should ideally
+gets its own config key and the inability to generate that config could still retain functionality for the rest of
+the project.
+
+### Separation of Traffic Classes
+
+Relay currently forwards all envelopes to the next relay in chain the same.  There are however benefits by being able
```

Comment:
open questions:

* Where should this split occur? I assume the simplest answer is to do the routing in pops, and only separate processing relays, but could it make sense to do it in customer relays already, so that pops are also segmented?